### PR TITLE
Capture Destination Metadata

### PR DESCRIPTION
**Pull request by prayansh**
- Add a new destination metadata plugin that captures `_metadata` info for Segment to handle delivery of events to cloud/device destinations
  - We keep the old format of `integrations` object being used by customers to control delivery
- Switch delivery of events from `/batch` to `/b` endpoint
- Add a `writeKey` field to the batch event file

### CODE REVIEW

File: core/src/main/java/com/segment/analytics/kotlin/core/HTTPClient.kt [hunk 1]
```diff
@@ -98,6 +97,7 @@ internal fun HttpURLConnection.createGetConnection(): Connection {
 
 internal fun HttpURLConnection.createPostConnection(): Connection {
     val outputStream: OutputStream
+    setRequestProperty("Content-Encoding", "gzip")
```
[the remark this answers is not in the document]
Comment (prayansh) [reply core/src/main/java/com/segment/analytics/kotlin/core/HTTPClient.kt:100]: just felt logically connected to the next line, that actually does the gzip encoding, so I moved it here

File: core/src/main/java/com/segment/analytics/kotlin/core/HTTPClient.kt [hunk 2]
```diff
@@ -27,19 +27,14 @@ class HTTPClient(private val writeKey: String) {
     }
 
     fun upload(apiHost: String): Connection {
-        val connection: HttpURLConnection = openConnection("https://$apiHost/batch")
+        val connection: HttpURLConnection = openConnection("https://$apiHost/b")
+        connection.setRequestProperty("Content-Type", "text/plain")
```
[the remark this answers is not in the document]
Comment (prayansh) [reply core/src/main/java/com/segment/analytics/kotlin/core/HTTPClient.kt:31]: `Content-Encoding` is being set to gzip, this is `Content-Type` which I believe has a different meaning

File: core/src/main/java/com/segment/analytics/kotlin/core/Events.kt [hunk 3]
```diff
@@ -79,6 +88,8 @@ sealed class BaseEvent {
     // the userId tied to the event
     abstract var userId: String
 
+    abstract var _metadata: DestinationMetadata
```
[the remark this answers is not in the document]
Comment (prayansh) [reply core/src/main/java/com/segment/analytics/kotlin/core/Events.kt:91]: Oh i didnt realize fields in an abstract class can be set as `lateinit`. That sounds like a good idea maybe. But I think that change goes beyond the scope of this PR. I will address those in another one after tho

File: core/src/main/java/com/segment/analytics/kotlin/core/HTTPClient.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (wenxi-zeng) [reply core/src/main/java/com/segment/analytics/kotlin/core/HTTPClient.kt:31]: ah. I read too fast. thought they are both `Content-Type`

**Review by wenxi-zeng — APPROVED** [review]

**Pull request merged**